cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -250,7 +260,53 @@ public void getAllApps(HttpRequest request, HttpResponder responder,
       .map(ApplicationRecord::new)
       .collect(Collectors.toList());
 
-    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(apps));
+    if (Optional.ofNullable(pageSize).orElse(0) != 0) {
+      PaginatedApplicationRecords result = getPaginatedResults(apps, pageToken, pageSize, orderBy,
+                                                               nameFilter);
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(result));
+    } else {
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(apps));
+    }
+  }
+
+  /**
+   * TODO: This is a temporary implementation, will be deprecated after
+   * https://cdap.atlassian.net/browse/CDAP-18603 is fully implemented.
+   * Returns paginated result for getApps, also includes nextPageToken
+   *
+   * @param apps the applications obtained from the Store
+   * @param pageToken the next pageToken to return the paginated results from
+   * @param pageSize the total number of application records to display on the result page
+   * @param orderBy  ASC for ascending order, DESC for descending order
+   * @param nameFilter filter applications whose name contains nameFilter
+   */
+  private PaginatedApplicationRecords getPaginatedResults(List<ApplicationRecord> apps,
+                                                           String pageToken, Integer pageSize,
+                                                           SortOrder orderBy, String nameFilter
+      ) {
+
+    if (pageToken != null && pageToken.equals("0")) {
+      return new PaginatedApplicationRecords(Collections.emptyList(), "0");
```

Comment:
What's logic here? I mean if anyone passes "0" as page token you get an recursive return of empty page with same token. Why do you need this if at all?

Reply:
As we discussed, I agree passing "0" introduces magic values.